Fix for big data passing

fenglixa opened the pull request:
**Which issue is resolved by this Pull Request:** 
Resolves #63
And related to #150

**Description of your changes:**

Big data should be read from files and written to files.

The paths for the input and output files are chosen by the system and are passed into the function (as strings).

Use the InputPath parameter annotation to tell the system that the function wants to consume the corresponding input data as a file. The system will download the data, write it to a local file and then pass the path of that file to the function.

Use the OutputPath parameter annotation to tell the system that the function wants to produce the corresponding output data as a file. The system will prepare and pass the path of a file where the function should write the output data. After the function exits, the system will upload the data to the storage system so that it can be passed to downstream components.

Notes: As we used 'workspaces' in tekton pipelines to handle the bigger data processing, the complier will generator the PVC definations,it needs the volume to store the data. User need to create volume manually, or enable dynamic volume provisioning, refer to the link of: https://kubernetes.io/docs/concepts/storage/dynamic-provisioning

- Implementation:

- [x] Index the PIPELINEs to understand how data is being passed and which inputs/outputs
       are connected to each other.
- [x] Search for direct data consumers in container/resource templates and some
       PIPELINE task attributes (e.g. conditions and loops) to find out which inputs
       are directly consumed as parameters/artifacts.
- [x] Propagate the consumption information upstream to all inputs/outputs all
       the way up to the data producers.
- [x] Convert the inputs, outputs based on how they're consumed downstream.
- [x] Use workspaces instead of result and params for bigger data passing.
- [x] Added workspaces to tasks, pipelines, pipelineruns, if the parmas is bigger data.
- [x] A PVC named pipelinerun name will be created if bigger data passing, as workspaces need to use it.
- [x] Document about 'How to use it'

fenglixa commented:
/cc @animeshsingh @ckadner @Tomcli 
Thanks very much!

Udiknedormin commented:
@fenglixa After merging and applying `generate_pipelinerun=True` flag (as per assertion's error message), some tasks get their parameters removed despite using them. Because of that, the task refers to a non-existing parameter and pipeline validation fails on "passing unnecessary input parameter".

I will provide you with details via internal communication channels.

Tomcli commented on sdk/python/kfp_tekton/compiler/_op_to_template.py:
```diff
@@ -451,16 +451,19 @@ def _op_to_template(op: BaseOp, enable_artifacts=False):
+        #                         volume_mount_step_template, volume_template, mounted_param_paths)
+        # template['spec']['steps'] = _prepend_steps([copy_inputs_step], template['spec']['steps'])
+        # _update_volumes(template, volume_mount_step_template, volume_template)
+        # Francis Todo: above info may need to be moved to bigger data passing logic
```
Do we still need to keep these commented code since you already moved them to the bigger data passing logic?

fenglixa replied:
Oh, Thank you. Tomcli. I forgot delete it.

Tomcli commented on sdk/python/kfp_tekton/compiler/_op_to_template.py:
```diff
@@ -69,7 +69,7 @@ def _get_resourceOp_template(op: BaseOp,
         'apiVersion': tekton_api_version,
         'kind': 'Task',
-        'metadata': {'name': name},
+        'metadata': {'name': name, 'labels': {'app': 'resourceOp'}},
```
@ckadner Does the new pipelineRun design allow adding labels to individual task?

fenglixa replied:
I added a labels for all the resourceOp tasks to distinguish resource and normal container. @ckadner, please let me know if not meet your requirements, or provide me a suggestion. Thanks! : )

ckadner replied:
no, with embedded `taskSpec` `metedata` is not allowed as far as I know

@afrittoli -- do you have any advice about enabling `metadata` for embedded `taskSpec`s?

fenglixa replied:
Drop the lables for resourceOp. Resolve this issue

Tomcli commented on sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py:
```diff
@@ -52,14 +53,615 @@ def fix_big_data_passing(workflow: List[Dict[Text, Any]]) -> List[Dict[Text, Any
+            "name": pr
+        },
+        "spec": {
+            "accessModes": ["ReadWriteOnce"],
```
With RWO, tasks in different k8s node won't able to write/mount at the same time. Is it the same behavior in Argo?

fenglixa replied:
Since I added the `runAfter` in the workspaces tasks. So there should be no situation of writing same file at same time

fenglixa replied:
OK， get it,  seems we should use RWX, but not RWO, will double check the behavior in Argo.

Tomcli commented on sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py:
```diff
@@ -52,14 +53,615 @@ def fix_big_data_passing(workflow: List[Dict[Text, Any]]) -> List[Dict[Text, Any
+# As we used workspaces in tekton pipelines which depends on it.
+# User need to create PV manually, or enable dynamic volume provisioning, refer to the link of:
+# https://kubernetes.io/docs/concepts/storage/dynamic-provisioning
+# TODO: Remove PVC if Tekton version > = 0.12, use 'volumeClaimTemplate' instead
```
The current design will use the same pvc for all the pipelinerun. Did Argo store them as "artifacts" in the original design?

fenglixa replied:
I difined different files for the bigger_data params,  I think we should reduce the pvc/pv number creating for one pipeline run, which mean reduce the effort for user of manual steps, there is no reason to create several pvcs for all the params.

fenglixa commented:
Thanks @Udiknedormin 's cooperation. Fixed the issue Michalina telled me, it's my code error. which was fixed by me just now.
Another issue we enconterred is an old issue which related to volumeMounts. I will provide another PR later to fix it.

animeshsingh commented:
Thanks @fenglixa - great job!!!! 

@Tomcli @Udiknedormin @ckadner any final comments, or is good to go?

ckadner commented on OWNERS:
```diff
@@ -2,6 +2,7 @@ approvers:
  - animeshsingh
  - ckadner
  - Tomcli
+ - fenglixa
```
@fenglixa @animeshsingh -- not debating merit, but I think that a promotion to become "committer" should be part of a separate PR

animeshsingh replied:
Definitely, you deserve to be an approver here @fenglixa. Let's just do it in a separate PR using the community norms, so that we can approve and celebrate!!!

fenglixa replied:
Thanks @ckadner, @animeshsingh , I will log separate PR for it.

ckadner commented on sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py:
```diff
@@ -52,14 +53,615 @@ def fix_big_data_passing(workflow: List[Dict[Text, Any]]) -> List[Dict[Text, Any
     return workflow
+
+
+# Replaced '_' to '-' to make the format in tekton unified for params, artifacts, outputs.
```
we have a general method that ["sanitizes" Kubernetes names](https://github.com/kubeflow/kfp-tekton/blob/2c68de4/sdk/python/kfp_tekton/compiler/_k8s_helper.py#L20) which should be used instead of `replace('_', '-')`

...which should have been invoked already by [`_sanitize_and_inject_artifact()`](https://github.com/kubeflow/kfp-tekton/blob/d806bc8/sdk/python/kfp_tekton/compiler/compiler.py#L534). If some names were missed it should be added there.

fenglixa replied:
Hi @ckadner ,
I tried to add the name and use _sanitize_and_inject_artifact() almost whole day yesterday, but failed, the major reasons:
1.  The params and artifacts cannot be converted before workflow and templates created. otherwise, some value will be mis-matched, it will impact to other features, e.g: loop, conditions. etc...
2. _sanitize_and_inject_artifact() is not suitable to use after workflow created, because a. `_sanitize_and_inject_artifact` will remove the symbol of "$,{,}, etc..", b.  `allow_capital_underscore` will convert  capital and underscore in the same time

Personally, I think I can provide another PR separately, try to unified the params in bigger data and _sanitize_and_inject_artifact. How do you think?

fenglixa commented on sdk/python/kfp_tekton/compiler/_op_to_template.py:
```diff
@@ -355,7 +355,7 @@ def _process_base_ops(op: BaseOp):
     # map param's (unsanitized pattern or serialized str pattern) -> input param var str
     map_to_tmpl_var = {
-        (param.pattern or str(param)): '$(inputs.params.%s)' % param.full_name  # Tekton change
+        (param.pattern or str(param)): '$(inputs.params.%s)' % param.full_name.replace('_', '-')  # Tekton change
```
I cannot use sanitize_k8s_name, since it will lower the letter if not allow_capital_underscore, which will not along to the outputs and other params.

fenglixa commented on sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py:
```diff
@@ -52,14 +53,615 @@ def fix_big_data_passing(workflow: List[Dict[Text, Any]]) -> List[Dict[Text, Any
     Implementation:
 
-    1. Index the DAGs to understand how data is being passed and which inputs/outputs
+    1. Index the PIPELINEs to understand how data is being passed and which inputs/outputs
```
Thanks @ckadner , Addressed

fenglixa commented on sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py:
```diff
@@ -52,14 +53,615 @@ def fix_big_data_passing(workflow: List[Dict[Text, Any]]) -> List[Dict[Text, Any
+    4. Convert the inputs, outputs based on how they're consumed downstream.
+    5. Use workspaces instead of result and params for bigger data passing.
+    6. Added workspaces to tasks, pipelines, pipelineruns, if the parmas is bigger data.
+    7. Convert "_" to "-" in the params and artifacts.
```
Thanks @ckadner , Yes, I tested, we have to covert _ to - for the params, since the results value didn't support it, and we already coverted it, this will lead to inputs and outputs inconsistency. This was tested out for my big data passing cases, and also reproduced for watson cases.

fenglixa commented on sdk/python/tests/compiler/testdata/artifact_location.yaml:
```diff
@@ -20,7 +20,7 @@ spec:
   - name: bucket
   - name: namespace
-  - name: secret_name
+  - name: secret-name
```
Thanks @ckadner, yes capital letters and underscore characters are support for parameter name, but not for results value, and we already coverted it in results value, this will lead to inputs and outputs inconsistency. It was tested out for my big data passing cases, and also reproduced for watson cases.

fenglixa commented:
> Any reason to use the term "big**ger**_data_passing" vs "big_data_passing" as used in KFP?

Good quetions, to avoid any confuse, I updated it all as `big_data_passing` which same as KFP. Thanks @ckadner

fenglixa commented:
Addressed the comments and fixed the issues for watson case, big data passing is working for old feature, big data, and also Watson pipelines now.
@Tomcli @ckadner @Udiknedormin , Is it OK for you now? thanks!

Udiknedormin commented:
@fenglixa Works fine on our side. Thank you!

animeshsingh commented:
/approve
